### PR TITLE
Add ToddCoxeter functionality

### PR DESCRIPTION
**Pull request by MTWhyte**

### CODE REVIEW

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/todd_coxeter.py]: Think this should be something else,  this is a `ToddCoxeter` not a `FroidurePin`

File: libsemigroups_cppyy/__init__.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/__init__.py]: This list of includes is in alphabetical order (well with the exception of report and schreier-sims being the wrong way around), can you please maintain/improve the alphabeticalisation here?

File: libsemigroups_cppyy/__init__.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/__init__.py]: This line is a duplicate of a line from earlier, delete

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [reply libsemigroups_cppyy/todd_coxeter.py]: In fact, why's this here at all, I'm guessing this is a mistake

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/todd_coxeter.py]: this could be `lambda x : "<ToddCoxeter object {0} generator{1} and {2} pair{3} at {4}>".format(x.nr_generators() if x.nr_generators() != undef else "-", "s"[: x.nr_generators() != 1], x.nr_generating_pairs(), "s"[: x.nr_generating_pairs() != 1], hex(id(x)))` maybe a bit more readable, also no `:` and include the memory address, as in FroidurePin

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/todd_coxeter.py]: This function could be removed and replaced with `undef` being defined in the outer scope, and then see my comment below.

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/todd_coxeter.py]: I think for consistency none of the exception messages should start with a capital letter, and ideally they'd be of the form:
"what is wrong, expected input, found input"
so in this example, it'd be `invalid argument, expected one of "right", "left" and "twosided", but found ` + t

File: libsemigroups_cppyy/todd_coxeter.py
Comment (james-d-mitchell) [inline libsemigroups_cppyy/todd_coxeter.py]: Same comment here

Comment (james-d-mitchell) [reply libsemigroups_cppyy/todd_coxeter.py]: as the one below about exceptions

File: tests/test_todd_coxeter.py
Comment (james-d-mitchell) [inline tests/test_todd_coxeter.py]: self.assertTrue

File: tests/test_todd_coxeter.py
Comment (james-d-mitchell) [inline tests/test_todd_coxeter.py]: self.assertFalse

**Review by james-d-mitchell — CHANGES_REQUESTED** [review]
Some changes required, and so more in the PR to your PR that I made earlier. Looks generally good though!

File: libsemigroups_cppyy/todd_coxeter.py
Comment (MTWhyte) [inline libsemigroups_cppyy/todd_coxeter.py]: This looks a little bit funny, but this is because using the `detail.wrap_overload_params` as opposed to `detail.wrap_overload_params_and_unwrap_return_value` gave me a return value looking like `(<ToddCoxeter object ... >, None)`. Changing the `__do_nothing` function in `detail.py` from returning `args` to returning `None` fixes this, but causes problems in places that I don't understand (the tests failed), and so I've just gone with this approach for now.

**Comment by james-d-mitchell** [comment]
I think your tests are failing probably for a similar reason as the one that hit my PR #46, I think I might know how to "fix" this, but it will involve some experimentation. Am I correct in thinking that the last test (the one which isn't terminating) is testing iterators?

**Comment by MTWhyte** [comment]
Yeah, the test that I've called `test_096` include a call to the iterator-based `normal_forms()`.

**Comment by james-d-mitchell** [comment]
Ok then this is the same cppyy bug that I "fixed" for my PR.

**Comment by MTWhyte** [comment]
I've commented out the problematic test and so the ToddCoxeter tests now pass. However, on the `libsemigroups-1.1.0` test, one of the tests in `test_fpsemi.py` now fails. As far as this PR is concerned, though, is that fine?

**Comment by james-d-mitchell** [comment]
@MTWhyte I'm afraid we'll have to fix the failing test before merging, I'm not sure if it's related to this PR or not, but either way, let's fix that first and then merge this PR.

**Comment by james-d-mitchell** [comment]
I'm happy with this now, although I just updated the branch after merging PR #52, and so will wait until the CI has finished. 

One very minor question: do the numbers in the test cases correspond to the numbers in the libsemigroups tests?

**Comment by MTWhyte** [comment]
> One very minor question: do the numbers in the test cases correspond to the numbers in the libsemigroups tests?

Yeah, they correspond, although I adapted them slightly.

**Review by james-d-mitchell — APPROVED** [review]
Awesome, going to merge now.

**Pull request merged**
